alacritty/src/renderer/shader.rs

```diff
@@ -60,23 +80,34 @@ impl Drop for ShaderProgram {
 struct Shader(GLuint);
 
 impl Shader {
-    fn new(kind: GLenum, source: &'static str) -> Result<Self, ShaderError> {
-        let len: [GLint; 1] = [source.len() as GLint];
+    fn new(
+        shader_version: ShaderVersion,
+        kind: GLenum,
+        source: &'static str,
+    ) -> Result<Self, ShaderError> {
+        let header = shader_version.shader_header();
+        let len: [GLint; 2] = [header.len() as GLint, source.len() as GLint];
+        let source = [header.as_ptr() as *const _, source.as_ptr() as *const _];
```

Comment:
You're assembling the shader here in OpenGL, any specific reason for that? I feel like the more intuitive approach would be to just append the two strings to each other?

Reply:
I mean there's a native way to append shaders, so I don't see why not? Instead of randomly appending rust strings.

Reply:
I'm just curious, I'm surprised it works like this at all? And surprises are usually a bad thing.

Reply:
The documentation clearly states that it concatenates all the sources, since api accepts array of sources. I'm not sure you should be surprised by documented feature.